src/appleseed/foundation/math/compressedunitvector.h

```diff
@@ -0,0 +1,179 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#pragma once
+
+// appleseed.foundation headers.
+#include "foundation/math/scalar.h"
+#include "foundation/math/vector.h"
+#include "foundation/platform/compiler.h"
+#include "foundation/platform/types.h"
+
+// Standard headers.
+#include <cmath>
+
+namespace foundation
+{
+
+//
+// The CompressedUnitVector class stores unit vectors
+// encoded as 32 bit octahedral unit vectors.
+//
+// Reference:
+//
+//   A Survey of Efficient Representations for Independent Unit Vectors.
+//   http://jcgt.org/published/0003/02/01/paper-lowres.pdf
+//
+
+class CompressedUnitVector
+{
+  public:
+    // Constructors.
+#if APPLESEED_COMPILER_CXX_DEFAULTED_FUNCTIONS
+    CompressedUnitVector() = default;     // leave uninitialized
+#else
+    CompressedUnitVector() {}             // leave uninitialized
+#endif
+
+    explicit CompressedUnitVector(const Vector3f& vec);
+
+    operator Vector3f() const;
+
+    bool operator==(const CompressedUnitVector& rhs) const;
+    bool operator!=(const CompressedUnitVector& rhs) const;
+
+  private:
+    static float sign_not_zero(const float v);
+    static int16 round16(const float f);
+    static int16 floor16(const float f);
+    static Vector3f oct_decode(const int16 bits[2]);
+
+    int16 m_bits[2];
+};
+
+
+//
+// CompressedUnitVector class implementation.
+//
+
+inline CompressedUnitVector::CompressedUnitVector(const Vector3f& vec)
+{
+    int16 projected[2];
+
+    const float inv_l1_norm = 1.0f / (std::abs(vec[0]) + std::abs(vec[1]) + std::abs(vec[2]));
+
+    if (vec[2] <= 0.0f)
+    {
+        projected[0] = floor16((1.0f - std::abs(vec[1] * inv_l1_norm)) * sign_not_zero(vec[0]));
+        projected[1] = floor16((1.0f - std::abs(vec[0] * inv_l1_norm)) * sign_not_zero(vec[1]));
+    }
+    else
+    {
+        projected[0] = floor16(vec[0] * inv_l1_norm);
+        projected[1] = floor16(vec[1] * inv_l1_norm);
+    }
+
+    int16 best_projected[2] = {0, 0};
+    float error = 0.0f;
+
+    unsigned int bits_y = projected[1];
```

Comment:
`projected` and `best_projected` are signed, but `bits_x` and `bits_y` aren't: intentional I suppose? If so, add `static_cast<>`.

Reply:
Swap the two lines (`bits_x` first, `bits_y` second).